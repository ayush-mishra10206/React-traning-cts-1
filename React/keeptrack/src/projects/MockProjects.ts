import { Project } from "./Project";

export const MOCK_PROJECTS = [
    new Project({
      id: 1,
      name: 'Perosn1',
      descp:'Descp1',
      imageUrl: '/assets/placeimg_500_300_arch4.jpg',
      contractTypeId: 3,
      contractSignedOn: '2013-08-04T22:39:41.473Z',
      budget: 54637,
      isActive: false,
    }),
    new Project({
      id: 2,
      name: 'Person2',
      descp:'Descp2',
      imageUrl: '/assets/placeimg_500_300_arch1.jpg',
      contractTypeId: 4,
      contractSignedOn: '2012-08-06T21:21:31.419Z',
      budget: 91638,
      isActive: true,
    }),
]