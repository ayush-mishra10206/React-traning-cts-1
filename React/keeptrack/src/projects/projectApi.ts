import { Project } from "./Project";

const baseUrl = 'http://localhost:3000';
const url = `${baseUrl}/projects`;




const delay = (ms: number) => {
    return (x: any): Promise<any> => {
        return new Promise((res) => setTimeout(() => res(x), ms))
    }
}

const translateStatusToErrorMessage = (status: number) => {
    switch (status) {
        case 401:
            return "login again";
        case 403:
            return "don't have permission"
        default:
            return "error in getting project"
    }
}

const checkStatus = (res: any) => {
    if (res.ok) {
        return res;
    }
    else {
        const httpError = {
            status: res.status,
            statusText: res.statusText,
            url: res.url
        }
        console.log('server error : ', httpError)
        let errorMessage = translateStatusToErrorMessage(httpError.status);
        throw new Error(errorMessage)
    }

}

const parseJSON = (response: Response) => {
    return response.json();
}

const convertToProjectModels = (data: any[]): Project[] => {
    let projects: Project[] = data.map(convertToProjectModel);
    return projects;
}

const convertToProjectModel = (item: any): Project => {
    return new Project(item);
}

export const projectApi = {
    get(page = 1, limit = 20) {
        return fetch(`${url}?_page=${page}&_limit=${limit}&_sort=name`)
            // .then(delay(600))
            .then(checkStatus)
            .then(parseJSON)
            .then(convertToProjectModels)
            .catch((error: TypeError) => {
                console.log('log client error ' , error);
                throw new Error(
                    'error retriveing data'
                );
            });
    },

    put(project: Project) {
        return fetch(`${url}/${project.id}`, {
            method: 'PUT',
            body: JSON.stringify(project),
            headers: {
                'Content-Type': 'application/json'
            }
        })
        .then(checkStatus)
        .then(parseJSON)
        .catch((error:TypeError)=>{
            console.log('log error ', error);
            throw new Error('error while updating')
        })
    }
}