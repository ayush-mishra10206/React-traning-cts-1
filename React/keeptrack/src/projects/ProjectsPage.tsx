import React, { Fragment, useEffect, useState } from 'react';
import { MOCK_PROJECTS } from './MockProjects';
import { Project } from './Project';
import { projectApi } from './projectApi';
import ProjectList from './ProjectList';

export default function ProjectPage() {
    // const [projects, setProjects] = useState<Project[]>(MOCK_PROJECTS)
    const [projects, setProjects] = useState<Project[]>([])
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);
    const [currentPage, setCurrentPage] = useState(1);


    useEffect(() => {
        const loadProjects = async () => {
            setLoading(true);
            try {
                const data = await projectApi.get(1);
                setError('');
                if (currentPage === 1) {
                    setProjects(data)
                }
                else {
                    setProjects([...projects, ...data])
                }
            }
            catch (e) {
                if (e instanceof Error) {
                    setError(e.message);
                }
            }
            finally {
                setLoading(false);
            }
        }
        loadProjects();
    }, [currentPage])

    const saveNewData = (project: Project) => {
        // let uptdProject = projects.map((prjt: Project) => {
        //     return prjt.id === project.id ? project : prjt;
        // });
        // setProjects(uptdProject);
        projectApi
            .put(project)
            .then((updatedProject) => {
                let updatedProjects = projects.map((p: Project) => {
                    return p.id === project.id ? new Project(updatedProject) : p;
                });
                setProjects(updatedProjects);
            })
            .catch((e) => {
                if (e instanceof Error) {
                    setError(e.message);
                }
            })
    }


    const handleMoreClick = () => {
        setCurrentPage((currentPage) => currentPage + 1);
    }


    return (
        <Fragment>
            <h1>Projects1</h1>
            {error && (
                <div className="row">
                    <div className="card large error">
                        <section>
                            <p>
                                <span className="icon-alert inverse "></span>
                                {error}
                            </p>
                        </section>
                    </div>
                </div>
            )}
            {!loading && !error && (
                <div className="row">
                    <div className="col-sm-12">
                        <div className="button-group fluid">
                            <button className="button default" onClick={handleMoreClick}>
                                More...
                            </button>
                        </div>
                    </div>
                </div>
            )}
            <ProjectList
                onClickedSave={saveNewData}
                projects={projects} />
            {loading && (
                <div className="center-page">
                    <span className="spinner primary"></span>
                    <p>Loading...</p>
                </div>
            )}
        </Fragment>
    )

}
