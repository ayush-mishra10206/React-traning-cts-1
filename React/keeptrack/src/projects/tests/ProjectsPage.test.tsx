import { render, screen } from "@testing-library/react"
import renderer from 'react-test-renderer';
import ProjectsPage from "../ProjectsPage";
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { notDeepEqual } from "assert";
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import { url as projectUrl } from "../projectApi";
import { MOCK_PROJECTS } from "../MockProjects";


const server = setupServer(
    rest.get(projectUrl,(req, res, ctx)=>{
        return res(ctx.json(MOCK_PROJECTS));
    })
)


describe('<ProjectsPage/>', () => {

    beforeAll(()=>server.listen());
    afterEach(()=>server.resetHandlers());
    afterAll(()=>server.close());

    test('working', () => {
        console.log('working without render')
        expect(screen).toBeDefined();
        // render(<ProjectPage/>)
    })

    test('testing loading component ', async () => {
        render(
            <MemoryRouter>
                <ProjectsPage />
            </MemoryRouter>)
            expect(screen.getByText(/loading/i)).toBeInTheDocument();
    })

    test('testing more button ', async()=>{
        render(
            <MemoryRouter>
                <ProjectsPage />
            </MemoryRouter>)
            expect(screen.queryByRole('button',{name:/more/i})).not.toBeInTheDocument()
    })

    test("should display project",async ()=>{
        render(
            <MemoryRouter>
                <ProjectsPage />
            </MemoryRouter>)
            expect(await screen.findAllByRole("img" )).toHaveLength(MOCK_PROJECTS.length);

    })



    xtest('snapshot', () => {
        const tree = renderer.create(<ProjectsPage />).toJSON();
        expect(tree).toMatchSnapshot();
    })
});