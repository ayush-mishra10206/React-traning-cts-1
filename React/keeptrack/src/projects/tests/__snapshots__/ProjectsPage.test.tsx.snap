// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectsPage/> snapshot 1`] = `
Array [
  <h1>
    Projects1
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-sm-12"
    >
      <div
        className="button-group fluid"
      >
        <button
          className="button default"
          onClick={[Function]}
        >
          More...
        </button>
      </div>
    </div>
  </div>,
  <div
    className="row"
  />,
]
`;
